proj/hps_accel/gateware/stream/test_buffer.py

```diff
@@ -79,3 +81,90 @@ def process():
+
+    def test(self):
+        DATA = [
+            # Format is: input x (valid, payload, ready),
+            #            input y (valid, payload, ready),
+            #            output (valid, payload, ready)
+            # Simultaneous inputs, passthrough, output constant ready
+            ((0, 0, 1), (0, 0, 1), (0, None, 1)),
+            ((1, 0xaa, 1), (1, 0xbb, 1), (1, 0xbbaa, 1)),
+            ((0, 0, 1), (0, 0, 1), (0, None, 1)),
+            # Simultaneous inputs, passthrough, output instantaneosly ready
+            ((0, 0, 1), (0, 0, 1), (0, None, 0)),
+            ((1, 0xcc, 1), (1, 0xdd, 1), (1, 0xddcc, 1)),
+            ((0, 0, 1), (0, 0, 1), (0, None, 0)),
+            # Simultaneous inputs, delay ready one cycle
+            ((0, 0, 1), (0, 0, 1), (0, None, 0)),
+            ((1, 0xee, 1), (1, 0xff, 1), (1, 0xffee, 0)),
+            ((0, 0, 1), (0, 0, 1), (1, 0xffee, 1)),
+            ((0, 0, 1), (0, 0, 1), (0, None, 0)),
+            # Simultaneous inputs, delay ready two cycle
+            ((0, 0, 1), (0, 0, 1), (0, None, 0)),
+            ((1, 0x77, 1), (1, 0x88, 1), (1, 0x8877, 0)),
+            ((0, 0, 0), (0, 0, 0), (1, 0x8877, 0)),
+            ((0, 0, 1), (0, 0, 1), (1, 0x8877, 1)),
+            ((0, 0, 1), (0, 0, 1), (0, None, 0)),
+            # Buffer a stream of simultaneous inputs
+            ((0, 0, 1), (0, 0, 1), (0, None, 0)),
+            ((1, 0x11, 1), (1, 0x22, 1), (1, 0x2211, 0)),
+            ((1, 0x33, 0), (1, 0x44, 0), (1, 0x2211, 0)),
+            ((1, 0x33, 0), (1, 0x44, 0), (1, 0x2211, 0)),
+            ((1, 0x33, 1), (1, 0x44, 1), (1, 0x2211, 1)),
+            ((1, 0x55, 1), (1, 0x66, 1), (1, 0x4433, 1)),
+            ((1, 0x77, 1), (1, 0x88, 1), (1, 0x6655, 1)),
+            ((0, 0, 0), (0, 0, 0), (1, 0x8877, 0)),
+            ((0, 0, 1), (0, 0, 1), (1, 0x8877, 1)),
+            ((0, 0, 1), (0, 0, 1), (0, None, 0)),
+            # One input arrives before the other, passthrough same cycle
+            ((1, 0xaa, 1), (0, 0, 1), (0, None, 0)),
+            ((0, 0, 1), (1, 0xbb, 1), (1, 0xbbaa, 1)),
+            ((0, 0, 1), (0, 0, 1), (0, None, 0)),
+            # One input arrives before the other, both buffered
+            ((1, 0xaa, 1), (0, 0, 1), (0, None, 0)),
+            ((0, 0, 0), (1, 0xbb, 1), (1, 0xbbaa, 0)),
+            ((0, 0, 1), (0, 0, 1), (1, 0xbbaa, 1)),
+            ((0, 0, 1), (0, 0, 1), (0, None, 0)),
+            # One input stream starts before the other
+            ((1, 0x11, 1), (0, 0, 1), (0, None, 0)),
+            ((1, 0x33, 0), (1, 0x22, 1), (1, 0x2211, 0)),
+            ((1, 0x33, 0), (1, 0x44, 0), (1, 0x2211, 0)),
+            ((1, 0x33, 1), (1, 0x44, 1), (1, 0x2211, 1)),
+            ((1, 0x55, 1), (1, 0x66, 1), (1, 0x4433, 1)),
+            ((1, 0x77, 1), (1, 0x88, 1), (1, 0x6655, 1)),
+            ((0, 0, 0), (0, 0, 0), (1, 0x8877, 0)),
+            ((0, 0, 1), (0, 0, 1), (1, 0x8877, 1)),
+            ((0, 0, 1), (0, 0, 1), (0, None, 0)),
```

Comment:
The length of this test case, and how difficult it was to write, is a good sign that the ConcatenatingBuffer might be doing too much or trying to be too clever :-)